File: exercise/java/day19/src/test/java/blog/ArticleBuilder.java
package blog;

import java.util.HashMap;

import static org.instancio.Instancio.create;

public class ArticleBuilder {
    public static final String AUTHOR = "Pablo Escobar";
    public static final String COMMENT_TEXT = "Amazing article !!!";
    private final HashMap<String, String> comments;

    public ArticleBuilder() {
        comments = new HashMap<>();
    }

    public static ArticleBuilder anArticle() {
        return new ArticleBuilder();
    }

    public ArticleBuilder commented() {
        this.comments.put(COMMENT_TEXT, AUTHOR);
        return this;
    }

    public Article build() {
        return comments.entrySet()
                .stream()
                .reduce(new Article(
                        create(String.class),
                        create(String.class)
                ), (a, e) -> a.addComment(e.getKey(), e.getValue()), (p, n) -> p);
    }
}

File: solution/ts/day19/tests/articleBuilder.ts
import {Article} from "../src/blog";
import {faker} from "@faker-js/faker";
import * as E from "fp-ts/Either";

export const Author = "Pablo Escobar";
export const CommentText = "Amazing article !!!";

export class ArticleBuilder {
    private comments: Record<string, string> = {};
    static anArticle = (): ArticleBuilder => new ArticleBuilder();

    commented(): ArticleBuilder {
        this.comments[CommentText] = Author;
        return this;
    }

    build(): Article {
        let article = Article.create(faker.lorem.sentence(), faker.lorem.paragraph());
        Object.entries(this.comments).forEach(([text, author]) => {
            let result = article.addComment(text, author);
            if (E.isRight(result)) article = result.right;
        });
        return article;
    }
}